Problem with importing new BOMs

### Current Behavior

when I build my projects on jenkins I currently use the API of Dependency Track to import the BOMs and create new projects for each version which worked great for the past months.
Now I suddenly get the following exception with 2 projects:

2024-07-09 13:15:43,664 INFO [BomUploadProcessingTask] Processing CycloneDX BOM uploaded to project: 361da822-885a-4ba3-915c-63c6280fd460
2024-07-09 13:15:45,199 ERROR [BomUploadProcessingTask] Error while processing bom
javax.jdo.JDOUserException: The query returned more than one instance BUT either unique is set to true or only aggregates are to be returned, so should have returned one result maximum
        at org.datanucleus.api.jdo.JDOAdapter.getJDOExceptionForNucleusException(JDOAdapter.java:698)
        at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:456)
        at org.datanucleus.api.jdo.JDOQuery.executeUnique(JDOQuery.java:360)
        at org.dependencytrack.persistence.LicenseQueryManager.getLicenseByIdOrName(LicenseQueryManager.java:102)
        at org.dependencytrack.persistence.QueryManager.getLicenseByIdOrName(QueryManager.java:624)
        at org.dependencytrack.parser.cyclonedx.util.ModelConverter.convert(ModelConverter.java:574)
        at org.dependencytrack.parser.cyclonedx.util.ModelConverter.convertComponents(ModelConverter.java:464)
        at org.dependencytrack.tasks.BomUploadProcessingTask.inform(BomUploadProcessingTask.java:157)
        at org.dependencytrack.tasks.BomUploadProcessingTaskV2.inform(BomUploadProcessingTaskV2.java:151)
        at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:110)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.datanucleus.store.query.QueryNotUniqueException: The query returned more than one instance BUT either unique is set to true or only aggregates are to be returned, so should have returned one result maximum
        at org.datanucleus.store.query.Query.executeQuery(Query.java:2039)
        at org.datanucleus.store.query.Query.executeWithMap(Query.java:1911)
        at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:437)
        ... 11 common frames omitted


### Steps to Reproduce

1.build 2 of my projects to get the exception

### Expected Behavior

Process the BOM

### Dependency-Track Version

4.11.5

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Database Server Version

14

### Browser

Google Chrome

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

--- a/src/main/java/org/dependencytrack/parser/cyclonedx/util/ModelConverter.java
+++ b/src/main/java/org/dependencytrack/parser/cyclonedx/util/ModelConverter.java
@@ -575,8 +575,8 @@ public class ModelConverter {
                         if (license != License.UNRESOLVED) {
                             component.setResolvedLicense(license);
                         } else {
-                            final License customLicense = qm.getCustomLicense(StringUtils.trimToNull(cycloneLicense.getName()));
-                            if (customLicense != null) {
+                            final License customLicense = qm.getCustomLicenseByName(StringUtils.trimToNull(cycloneLicense.getName()));
+                            if (customLicense != License.UNRESOLVED) {
                                 component.setResolvedLicense(customLicense);
                             }
                         }

--- a/src/main/java/org/dependencytrack/persistence/LicenseQueryManager.java
+++ b/src/main/java/org/dependencytrack/persistence/LicenseQueryManager.java
@@ -98,6 +98,8 @@ final class LicenseQueryManager extends QueryManager implements IQueryManager {
         final Query<License> query = pm.newQuery(License.class);
         query.setFilter("licenseId == :licenseIdOrName || name == :licenseIdOrName");
         query.setNamedParameters(Map.of("licenseIdOrName", licenseIdOrName));
+        query.setOrdering("licenseId asc"); // Ensure result is consistent.
+        query.setRange(0, 1); // Multiple licenses can have the same name; Pick the first one.
         try {
             final License license = query.executeUnique();
             return license != null ? license : License.UNRESOLVED;
@@ -116,6 +118,21 @@ final class LicenseQueryManager extends QueryManager implements IQueryManager {
         query.getFetchPlan().addGroup(License.FetchGroup.ALL.name());
         query.setRange(0, 1);
         return singleResult(query.execute(licenseName));
+    }
+
+    @Override
+    public License getCustomLicenseByName(final String licenseName) {
+        final Query<License> query = pm.newQuery(License.class);
+        query.setFilter("name == :name && customLicense == true");
+        query.setParameters(licenseName);
+        query.setOrdering("licenseId asc"); // Ensure result is consistent.
+        query.setRange(0, 1); // Multiple licenses can have the same name; Pick the first one.
+        try {
+            final License license = query.executeUnique();
+            return license != null ? license : License.UNRESOLVED;
+        } finally {
+            query.closeAll();
+        }
     }
 
     /**

--- a/src/main/java/org/dependencytrack/persistence/QueryManager.java
+++ b/src/main/java/org/dependencytrack/persistence/QueryManager.java
@@ -628,6 +628,10 @@ public class QueryManager extends AlpineQueryManager {
         return getLicenseQueryManager().getCustomLicense(licenseName);
     }
 
+    public License getCustomLicenseByName(final String licenseName) {
+        return getLicenseQueryManager().getCustomLicenseByName(licenseName);
+    }
+
     public License synchronizeLicense(License license, boolean commitIndex) {
         return getLicenseQueryManager().synchronizeLicense(license, commitIndex);
     }

--- a/src/main/java/org/dependencytrack/tasks/BomUploadProcessingTaskV2.java
+++ b/src/main/java/org/dependencytrack/tasks/BomUploadProcessingTaskV2.java
@@ -689,8 +689,7 @@ public class BomUploadProcessingTaskV2 implements Subscriber {
                     break;
                 }
 
-                final License resolvedCustomLicense = customLicenseCache.computeIfAbsent(licenseCandidate.getName(),
-                        licenseName -> resolveCustomLicense(qm, licenseName));
+                final License resolvedCustomLicense = customLicenseCache.computeIfAbsent(licenseCandidate.getName(), qm::getCustomLicenseByName);
                 if (resolvedCustomLicense != License.UNRESOLVED) {
                     component.setResolvedLicense(resolvedCustomLicense);
                     component.setLicenseUrl(trimToNull(licenseCandidate.getUrl()));
@@ -709,18 +708,6 @@ public class BomUploadProcessingTaskV2 implements Subscriber {
                         component.setLicense(trim(license.getName()));
                         component.setLicenseUrl(trimToNull(license.getUrl()));
                     });
-        }
-    }
-
-    private static License resolveCustomLicense(final QueryManager qm, final String licenseName) {
-        final Query<License> query = qm.getPersistenceManager().newQuery(License.class);
-        query.setFilter("name == :name && customLicense == true");
-        query.setParameters(licenseName);
-        try {
-            final License license = query.executeUnique();
-            return license != null ? license : License.UNRESOLVED;
-        } finally {
-            query.closeAll();
         }
     }
 

--- a/src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java
+++ b/src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java
@@ -1277,6 +1277,55 @@ public class BomUploadProcessingTaskTest extends PersistenceCapableTest {
         }
     }
 
+    @Test // https://github.com/DependencyTrack/dependency-track/issues/3957
+    public void informIssue3957Test() {
+        final var licenseA = new License();
+        licenseA.setLicenseId("GPL-1.0");
+        licenseA.setName("GNU General Public License v1.0 only");
+        qm.persist(licenseA);
+
+        final var licenseB = new License();
+        licenseB.setLicenseId("GPL-1.0-only");
+        licenseB.setName("GNU General Public License v1.0 only");
+        qm.persist(licenseB);
+
+        final var project = new Project();
+        project.setName("acme-license-app");
+        qm.persist(project);
+
+        final byte[] bomBytes = """
+                {
+                  "bomFormat": "CycloneDX",
+                  "specVersion": "1.4",
+                  "serialNumber": "urn:uuid:3e671687-395b-41f5-a30f-a58921a69b80",
+                  "version": 1,
+                  "components": [
+                    {
+                      "type": "library",
+                      "name": "acme-lib-x",
+                      "licenses": [
+                        {
+                          "license": {
+                            "name": "GNU General Public License v1.0 only"
+                          }
+                        }
+                      ]
+                    }
+                  ]
+                }
+                """.getBytes(StandardCharsets.UTF_8);
+
+        final var bomUploadEvent = new BomUploadEvent(qm.detach(Project.class, project.getId()), bomBytes);
+        new BomUploadProcessingTaskV2().inform(bomUploadEvent);
+        awaitBomProcessedNotification(bomUploadEvent);
+
+        qm.getPersistenceManager().evictAll();
+        assertThat(qm.getAllComponents(project)).satisfiesExactly(component -> {
+            assertThat(component.getResolvedLicense()).isNotNull();
+            assertThat(component.getResolvedLicense().getLicenseId()).isEqualTo("GPL-1.0");
+        });
+    }
+
     private void awaitBomProcessedNotification(final BomUploadEvent bomUploadEvent) {
         try {
             await("BOM Processed Notification")

